### PR TITLE
Added ReinforceGamma and fixed dpnnbigtest.Reinforce()

### PR DESCRIPTION
**Pull request by ssampang**
While I've edited dpnnbigtest.Reinforce() to conform to the current implementation of VRClassReward, I still don't understand it fully. The tests pass, but I'm not sure if comparing the reward and baseReward [here](https://github.com/ssampang/dpnn/blob/6857c44707ccfa2f1fd433a56d3c962ac3782e55/test/test.lua#L1672-L1674) is valid, or if its an artifact of how VRClassReward used to work. If you have time, it'd be great if you could give me a quick description of what's going on in the test, and/or compare it to the test I described in #52.

### CODE REVIEW

**Comment by nicholas-leonard** [comment]
@ssampang This PR looks awesome. You did a great job with the buffers and unit test. And thanks a million for fixing `dpnnbigtest.Reinforce`. To answer your question, yes your test is valid : 

``` lua
            if reward*0.7 > baseReward then
               converged = true
               break
            end
```

Basically, the above poses the question : is the baseReward lesser than 70% of reward after training? i.e. did the reward increase sufficiently to demonstrate that the Reinforce module has produced valid gradients? Since the VRClassReward criterion does not produce gradInputs for the output layer, i.e. Linear + Softmax, then the only way for the model to learn to maximize reward is through the Reinforce module's generated gradInputs. Which are back-propagated to the input layer (first Linear), which is the only one that can adapt its parameters.

Only one thing left : documentation for the README.

**Comment by ssampang** [comment]
I guess I haven't faced this issue yet because I haven't observed the behavior of my model when the `stochastic` flag is off, but should I be outputting the mean of the distribution when the `stochastic` flag is set? The gamma distribution is parametrized by a `shape` and a `scale` variable, and the mean of the distribution is defined as `shape*scale`. Right now, I'm just outputting the shape, but I think it makes more sense to output the mean, and it would be a simple fix.

**Comment by ssampang** [comment]
I changed the output to equal the mean on `stochastic=false`. It passes all the tests, so I pushed the changes, but I can revert if you think its necessary.

**Comment by nicholas-leonard** [comment]
@ssampang Looks perfect. Great PR. Thank you!
